pkg/agent/packetcapture/packetcapture_controller_test.go

```diff
@@ -0,0 +1,604 @@
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions"
+	"antrea.io/antrea/pkg/util/k8s"
+)
+
+var (
+	pod1IPv4 = "192.168.10.10"
+	pod2IPv4 = "192.168.11.10"
+
+	ipv6       = "2001:db8::68"
+	pod1MAC, _ = net.ParseMAC("aa:bb:cc:dd:ee:0f")
+	pod2MAC, _ = net.ParseMAC("aa:bb:cc:dd:ee:00")
+	ofPortPod1 = uint32(1)
+	ofPortPod2 = uint32(2)
+
+	icmpProto    = "ICMP"
+	udpProto     = "UDP"
+	shortTimeout = uint16(1)
+
+	pod1 = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod-1",
+			Namespace: "default",
+		},
+		Status: v1.PodStatus{
+			PodIP: pod1IPv4,
+		},
+	}
+	pod2 = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod-2",
+			Namespace: "default",
+		},
+		Status: v1.PodStatus{
+			PodIP: pod2IPv4,
+		},
+	}
+	pod3 = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod-3",
+			Namespace: "default",
+		},
+	}
+
+	secret1 = v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fileServerAuthSecretName,
+			Namespace: "kube-system",
+		},
+		Data: map[string][]byte{
+			"username": []byte("username"),
+			"password": []byte("password"),
+		},
+	}
+)
+
+func generateTestSecret() *v1.Secret {
+	return &v1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "AAA",
+			Namespace: "default",
+		},
+		Data: map[string][]byte{
+			"username": []byte("AAA"),
+			"password": []byte("BBBCCC"),
+		},
+	}
+}
+
+func genTestCR(name string, num int32) *crdv1alpha1.PacketCapture {
+	result := &crdv1alpha1.PacketCapture{
+		ObjectMeta: metav1.ObjectMeta{Name: name, UID: types.UID(fmt.Sprintf("uid-%s", name))},
+		Spec: crdv1alpha1.PacketCaptureSpec{
+			Source: crdv1alpha1.Source{
+				Pod: &crdv1alpha1.PodReference{
+					Namespace: pod1.Namespace,
+					Name:      pod1.Name,
+				},
+			},
+			Destination: crdv1alpha1.Destination{
+				Pod: &crdv1alpha1.PodReference{
+					Namespace: pod2.Namespace,
+					Name:      pod2.Name,
+				},
+			},
+			CaptureConfig: crdv1alpha1.CaptureConfig{
+				FirstN: &crdv1alpha1.PacketCaptureFirstNConfig{
+					Number: num,
+				},
+			},
+			Packet: &crdv1alpha1.Packet{
+				Protocol: &icmpProto,
+			},
+			FileServer: &crdv1alpha1.PacketCaptureFileServer{
+				URL: "sftp://127.0.0.1:22/aaa",
+			},
+		},
+	}
+	return result
+}
+
+type testUploader struct {
+	url      string
+	fileName string
+}
+
+func (uploader *testUploader) Upload(url string, fileName string, config *ssh.ClientConfig, outputFile afero.File) error {
+	if url != uploader.url {
+		return fmt.Errorf("expected url: %s for uploader, got: %s", uploader.url, url)
+	}
+	if fileName != uploader.fileName {
+		return fmt.Errorf("expected filename: %s for uploader, got: %s", uploader.fileName, fileName)
+	}
+	return nil
+}
+
+type testCapture struct {
+}
+
+func (p *testCapture) Capture(ctx context.Context, device string, srcIP, dstIP net.IP, packet *crdv1alpha1.Packet) (chan gopacket.Packet, error) {
```

Comment:
I'm aware of the mock solution.  But since the pcap interface provide such abstraction so we can use pcap file(data) to simulate the real traffic. Isn't this better?

cc @tnqn  Any thoughts?

Reply:
As discussed, crafe packet  struct + channel is better for this case. updated.

Reply:
OK.